File: dining-savages/ds.swift
//
//  main.swift
//  SwiftConcurrency
//
//  Created by TonyNguyen on 9/19/18.
//  Copyright © 2018 TonyNguyen. All rights reserved.
//

import Foundation

class Pot {
  var size: Int
  var servings: Int

  init(size: Int) {
    self.size = size
    self.servings = size
  }

  func fill() {
    self.servings = size
  }

  func isEmpty() -> Bool {
    return self.servings == 0
  }
}

let savages = DispatchQueue(label: "savages")
let pot = Pot(size: 10)
let group = DispatchGroup()

for i in 1...100 {
  group.enter()
  DispatchQueue.global().async {
    savages.sync {
      if pot.isEmpty() {
        print("Pot is empty, filling to \(pot.size)")
        pot.fill()
      }
      print("Current servings: \(pot.servings) - Savage \(i) is eating")
      pot.servings -= 1
      group.leave()
    }
  }
}
group.wait()
print("Done")

File: dining-savages/ds.go
package main

import (
	"fmt"
	"sync"
	"time"
)

var (
	potEmpty = make(chan bool)
	potFull  = make(chan bool)
	servings = 100
	mu       = sync.Mutex{}
	wg       = sync.WaitGroup{}
)

func cook() {
	for {
		<-potEmpty
		servings = 100
		potFull <- true
	}
}

func eat(savage int) {
	mu.Lock()
	if servings == 0 {
		potEmpty <- true
		// fmt.Printf("Savage %d, pot is empty, filling", savage)
		<-potFull
	}
	// fmt.Printf("Current servings: %d, savage %d is eating\n", servings, savage)
	servings--
	mu.Unlock()
	wg.Done()
}

func main() {
	start := time.Now()
	wg.Add(1000)
	go cook()
	for i := 1; i <= 1000; i++ {
		go eat(i)
	}
	wg.Wait()
	elapsed := time.Since(start)
	fmt.Printf("Total time %s", elapsed)
}
